sw/airborne/firmwares/rotorcraft/guidance/guidance_v.c

```diff
@@ -481,8 +481,18 @@ static void run_hover_loop(bool in_flight)
 bool guidance_v_set_guided_z(float z)
 {
   if (guidance_v_mode == GUIDANCE_V_MODE_GUIDED) {
+    /* disable vertical velocity setpoints */
     guidance_v_guided_vel_enabled = false;
+
+    /* set altitude setpoint */
     guidance_v_z_sp = POS_BFP_OF_REAL(z);
+
+    /* reset speed setting */
+    guidance_v_zd_sp = 0;
+
+    /* reset guidance reference */
+    guidance_v_z_sum_err = 0;
+    GuidanceVSetRef(guidance_v_z_sp, guidance_v_zd_sp, 0);
```

Comment:
You are actually completely circumventing the reference here as you set to the target setpoint...
IMHO the thing it makes sense to manually set the reference to is the current position/velocity
```
GuidanceVSetRef(stateGetPositionNed_i()->z, stateGetSpeedNed_i()->z, 0);
```
If it makes sense to do that in every set_guided_x function or not depends on how you want those command to behave in the end... in any case we should clearly state this in the function description.
Doing that in init makes no sense as you neither have a setpoint nor the current position yet...